From 7a573adb9d76d2ee02e5d53cae56e0b0e9b568dc Mon Sep 17 00:00:00 2001
From: Ken Zhu <guigenz@codeaurora.org>
Date: Wed, 10 Oct 2018 15:23:10 -0700
Subject: [PATCH] map: support direct communication among CPEs with the same
 ipv4 address

This change makes the CPEs with the same ipv4 address could communicate
with each other. All the packets with the destination ports belonging to
its own port set will be terminated in the CPE. Other packets continue to
be forwarded.

Change-Id: I91279a58fd0b95f3720b84ea8cd47f9d163ef5c2
Signed-off-by: Ken Zhu <guigenz@codeaurora.org>
(cherry picked from commit 06cb0295f19715cdc01f1b292ec95a7c84a41491)
---
 package/network/ipv6/map/files/map.sh | 24 ++++++++++++++++++++----
 1 file changed, 20 insertions(+), 4 deletions(-)

diff --git a/package/network/ipv6/map/files/map.sh b/package/network/ipv6/map/files/map.sh
index abd4f631feae..f56fa54ad4c9 100755
--- a/package/network/ipv6/map/files/map.sh
+++ b/package/network/ipv6/map/files/map.sh
@@ -147,7 +147,8 @@ proto_map_setup() {
 	      json_add_string family inet
 	      json_add_string snat_ip $(eval "echo \$RULE_${k}_IPV4ADDR")
 	    json_close_object
-	   else
+	  else
+	    network_get_device ifname "lan"
 	    for portset in $(eval "echo \$RULE_${k}_PORTSETS"); do
               for proto in icmp tcp udp; do
 	        json_add_object ""
@@ -161,9 +162,25 @@ proto_map_setup() {
 	        json_close_object
               done
 	    done
-	   fi
+	    for portset in $(eval "echo \$RULE_${k}_PORTSETS"); do
+	      json_add_object ""
+	        json_add_string type rule
+		json_add_string family inet
+		json_add_string set_mark 0x100
+		json_add_string dest_ip $(eval "echo \$RULE_${k}_IPV4ADDR")
+		json_add_string proto tcpudp
+		json_add_string src "lan"
+		json_add_string device "$ifname"
+		json_add_string dest_port "$portset"
+		json_add_string target MARK
+              json_close_object
+	    done
+	    ip rule add to $(eval "echo \$RULE_${k}_IPV4ADDR") iif $ifname fwmark 0x100/0x100 table local
+	    ip rule add to $(eval "echo \$RULE_${k}_IPV4ADDR") iif $ifname table main
 	  fi
-	  if [ "$maptype" = "map-t" ]; then
+	fi
+
+	if [ "$maptype" = "map-t" ]; then
 		[ -z "$zone" ] && zone=$(fw3 -q network $iface 2>/dev/null)
 
 		[ -n "$zone" ] && {
@@ -222,7 +239,6 @@ proto_map_teardown() {
 	[ -z "$maptype" ] && maptype="map-e"
 
 	case "$maptype" in
-		"map-e"|"lw4o6") ifdown "${cfg}_" ;;
 		"map-t") [ -f "/proc/net/nat46/control" ] && echo del $link > /proc/net/nat46/control ;;
 	esac
 
-- 
2.34.1

